File: crackup/classes/Crackup.php
<?php
/**
 * Crackup is a static class providing a namespace for Crackup's core
 * functionality.
 * 
 * @static
 * @author Ryan Grove <ryan@wonko.com>
 * @package Crackup
 */
class Crackup {
  // -- Public Static Variables ------------------------------------------------
  public static $remote      = '';
  public static $remoteFiles = array();
  public static $passphrase  = '';
  public static $local       = array();
  public static $localFiles  = array();
  
  // -- Public Static Methods --------------------------------------------------
  
  /**
   * Prints the specified debugging message if verbose mode is enabled.
   * 
   * @param String $message message to print
   */
  public static function debug($message) {
    if (defined('VERBOSE')) {
      echo "--> ".$message."\n";
    }
  }
  
  /**
   * Decrypts the specified local file using GPG.
   * 
   * @param String $infile filename of the file to decrypt
   * @param String $outfile filename of the file to decrypt to
   */
  public static function decryptFile($infile, $outfile) {
    $gpgCommand = strtr(GPG_DECRYPT, array(
      ':input_file'  => escapeshellarg($infile),
      ':output_file' => escapeshellarg($outfile),
      ':passphrase'  => escapeshellarg(Crackup::$passphrase)
    ));
    
    @shell_exec($gpgCommand);
  }
  
  /**
   * Encrypts the specified local file using GPG.
   *
   * @param String $infile filename of the file to encrypt
   * @param String $outfile filename of the file to encrypt to
   */
  public static function encryptFile($infile, $outfile) {
    $gpgCommand = strtr(GPG_ENCRYPT, array(
      ':input_file'  => escapeshellarg($infile),
      ':output_file' => escapeshellarg($outfile),
      ':passphrase'  => escapeshellarg(Crackup::$passphrase)
    ));
    
    @shell_exec($gpgCommand);
  }
  
  /**
   * Prints the specified error message to stderr and exists with an error code.
   *
   * @param String $message message to print
   */
  public static function error($message) {
    file_put_contents('php://stderr', "Error: ".$message."\n");
    exit(1);
  }
  
  /**
   * Searches the remote file index for a file whose local filename matches the
   * name specified and returns it if found, or <em>false</em> otherwise.
   *
   * @param String $filename local filename to search for
   * @return mixed
   */
  public static function findRemoteFile($filename) {
    $filename = rtrim($filename, '/');
    
    foreach(Crackup::$remoteFiles as $file) {
      if ($file->getName() == $filename) {
        return $file;
      }
      
      if ($file instanceof CrackupDirectory) {
        if ($child = $file->find($filename)) {
          return $child;
        }
      }
    }
    
    return false;
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing the files and
   * directories on the local system in the locations specified by the array of
   * file patterns in Crackup::$local.
   *
   * @return Array
   * @see getRemoteFiles()
   */
  public static function getLocalFiles() {
    $localFiles = array();
    
    foreach(Crackup::$local as $pattern) {
      if (false === ($filenames = glob($pattern, GLOB_NOSORT))) {
        continue;
      }
      
      foreach($filenames as $filename) {
        $filename = rtrim($filename, '/');
        
        if (isset($localFiles[$filename])) {
          continue;
        }
        
        if (is_dir($filename)) {
          Crackup::debug($filename);
          $localFiles[$filename] = new CrackupDirectory($filename);
        }
        elseif (is_file($filename)) {
          Crackup::debug($filename);
          $localFiles[$filename] = new CrackupFile($filename);
        }
      }
    }
    
    return $localFiles;
  }
  
  /**
   * Returns an array of CrackupFileSystemObject objects present at the remote
   * location.
   *
   * @return Array
   * @see getLocalFiles()
   */
  public static function getRemoteFiles() {
    if (!($fileList = @file_get_contents(self::$remote.'/.crackup_index'))) {
      return array();
    }
    
    if (defined('CRACKUP_NOGPG')) {
      $fileList = @unserialize($fileList);
    }
    else {
      $tmpName1 = self::getTempFile();
      $tmpName2 = self::getTempFile();
      
      if (!@file_put_contents($tmpName1, $fileList)) {
        @unlink($tmpName1);
        @unlink($tmpName2);
        Crackup::error('Unable to write to temporary directory');
      }
      
      self::decryptFile($tmpName1, $tmpName2);
      
      $fileList = @unserialize(@file_get_contents($tmpName2));
      
      @unlink($tmpName1);
      @unlink($tmpName2);
    }
    
    if (is_array($fileList)) {
      return $fileList;
    }
    
    return array();
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing files and
   * directories that exist on the destination but no longer exist on the local
   * machine.
   *
   * @param Array $localFiles array of local CrackupFileSystemObject objects
   * @param Array $remoteFiles array of remote CrackupFileSystemObject objects
   * @return Array removed CrackupFileSystemObject objects
   * @see getUpdatedFiles
   */
  public static function getRemovedFiles($localFiles, $remoteFiles) {
    $removed = array();
    
    foreach($remoteFiles as $name => $remoteFile) {
      if (!isset($localFiles[$name])) {
        $removed[] = $remoteFile;
        continue;
      }
      
      $localFile = $localFiles[$name];
      
      if (($remoteFile instanceof CrackupDirectory) && 
          ($localFile instanceof CrackupDirectory)) {
            
        $removed = array_merge($removed, self::getRemovedFiles(
            $localFile->getChildren(), $remoteFile->getChildren()));
      }
    }
    
    return $removed;
  }
  
  /**
   * Returns a PHP stream context containing settings to be used for stream
   * operations.
   * 
   * @return stream context
   */
  public static function getStreamContext() {
    return stream_context_create(
      array(
        'ftp' => array(
          'overwrite' => true,
        ),
      )
    );
  }
  
  /**
   * Generates a unique temporary filename.
   */
  public static function getTempFile() {
    return TEMP_DIR.'/'.uniqid('.crackup_temp');
  }
  
  /**
   * Gets an array of CrackupFileSystemObject objects representing files and
   * directories that are new or have been modified on the local machine and
   * need to be updated at the destination location.
   *
   * @param Array $localFiles array of local CrackupFileSystemObject objects
   * @param Array $remoteFiles array of remote CrackupFileSystemObject objects
   * @return Array updated CrackupFileSystemObject objects
   * @see getRemovedFiles()
   */
  public static function getUpdatedFiles($localFiles, $remoteFiles) {
    $updated = array();
    
    foreach($localFiles as $name => $localFile) {
      if (!isset($remoteFiles[$name])) {
        $updated[] = $localFile;
        continue;
      }
      
      $remoteFile = $remoteFiles[$name];
      
      if (($localFile instanceof CrackupDirectory) &&
          ($remoteFile instanceof CrackupDirectory)) {

        $updated = array_merge($updated, self::getUpdatedFiles(
            $localFile->getChildren(), $remoteFile->getChildren()));
      }
      elseif (($localFile instanceof CrackupFile) &&
          ($remoteFile instanceof CrackupFile)) {

        if ($localFile->getFileHash() != $remoteFile->getFileHash()) {
          $updated[] = $localFile;
        }
      }
    }

    return $updated;
  }
  
  /**
   * Displays a message and prompts for user input.
   *
   * @param String $message message to display
   * @return String user input
   */
  public static function prompt($message) {
    echo $message.': ';
    return trim(fgets(STDIN));
  }
  
  /**
   * Deletes the specified files/directories from the remote location.
   *
   * @param Array $files array of CrackupFileSystemObject objects to delete
   * @see update()
   */
  public static function remove($files) {
    foreach($files as $file) {
      if ($file instanceof CrackupDirectory) {
        self::remove($file);
      }
      elseif ($file instanceof CrackupFile) {
        Crackup::debug($file->getName());
        
        if (false === @unlink(self::$remote.'/crackup_'.$file->getNameHash())) {
          Crackup::error('Unable to remove "'.$file->getName().'" from '.
              'destination');
        }
      }
    }
  }
  
  /**
   * Uploads the specified files/directories to the remote location.
   *
   * @param Array $files array of CrackupFileSystemObject objects to update
   * @see remove()
   */
  public static function update($files) {
    foreach($files as $file) {
      if ($file instanceof CrackupDirectory) {
        self::update($file);
      }
      elseif ($file instanceof CrackupFile) {
        Crackup::debug($file->getName());
        
        $remoteFile = self::$remote.'/crackup_'.$file->getNameHash();
        
        // We have to manually delete the remote file if it already exists,
        // since some stream protocols don't allow overwriting by default.
        // FIXME: For some reason, file_exists returns false even when the file exists on an FTP server. WTF?
        if (@file_exists($remoteFile)) {
          if (false === @unlink($remoteFile)) {
            Crackup::error('Unable to remove "'.$file->getName().'" from '.
                'destination prior to update');
          }
        }
        
        if (defined('CRACKUP_NOGPG')) {
          if (false === @copy($file->getName(), $remoteFile)) {
            Crackup::error('Unable to update "'.$file->getName().'" at '.
                'destination');
          }
        }
        else {
          // Create a temporary local file and compress/encrypt it with GPG.
          $tmpName = self::getTempFile();

          self::encryptFile($file->getName(), $tmpName);          
          
          // Copy the compressed/encrypted temporary file to the destination
          // location, then delete the local copy.
          if (false === @copy($tmpName, $remoteFile)) {
            @unlink($tmpName);
            Crackup::error('Unable to update "'.$file->getName().'" at '.
                'destination');
          }
          
          @unlink($tmpName);
        }
      }
    }
  }
  
  /**
   * Brings the remote file index up to date with the local one.
   */
  public static function updateRemoteIndex() {
    $index      = serialize(Crackup::$localFiles);
    $remoteFile = Crackup::$remote.'/.crackup_index';
    
    if (defined('CRACKUP_NOGPG')) {
      while (!@file_put_contents($remoteFile, $index, null,
          Crackup::getStreamContext())) {

        $tryAgain = Crackup::prompt('Unable to update remote index. Try '.
            'again? (y/n)');

        if (strtolower($tryAgain) != 'y') {
          exit(-1);
        }
      }
    }
    else {
      $tmpName1 = Crackup::getTempFile();
      $tmpName2 = Crackup::getTempFile();
      
      while (!@file_put_contents($tmpName1, $index)) {
        $tryAgain = Crackup::prompt('Unable to write to temporary directory. '.
            'Try again? (y/n)');
        
        if (strtolower($tryAgain) != 'y') {
          @unlink($tmpName1);
          @unlink($tmpName2);
          exit(-1);
        }
      }
      
      self::encryptFile($tmpName1, $tmpName2);
      
      if (@file_exists($remoteFile)) {
        @unlink($remoteFile);
      }
      
      while (!@copy($tmpName2, $remoteFile)) {
        $tryAgain = Crackup::prompt('Unable to update remote index. Try '.
            'again? (y/n)');
        
        if (strtolower($tryAgain) != 'y') {
          @unlink($tmpName1);
          @unlink($tmpName2);
          exit(-1);
        }
      }
      
      @unlink($tmpName1);
      @unlink($tmpName2);
    }
  }
}
?>

File: crackup-sqlite3/lib/crackup.rb
ENV['PATH'] = "#{File.dirname(__FILE__)};#{ENV['PATH']}"

require 'crackup/errors'
require 'crackup/dirobject'
require 'crackup/driver'
require 'crackup/fileobject'
require 'crackup/index'
require 'tempfile'
require 'yaml'
require 'zlib'

module Crackup

  GPG_DECRYPT = 'echo :passphrase | gpg --batch --quiet --no-tty --no-secmem-warning --cipher-algo aes256 --compress-algo bzip2 --passphrase-fd 0 --output :output_file :input_file'
  GPG_ENCRYPT = 'echo :passphrase | gpg --batch --quiet --no-tty --no-secmem-warning --cipher-algo aes256 --compress-algo bzip2 --passphrase-fd 0 --output :output_file --symmetric :input_file'
  
  attr_accessor :driver, :local_files, :options, :remote_files
  
  # Reads <em>infile</em> and compresses it to <em>outfile</em> using zlib
  # compression.
  def self.compress_file(infile, outfile)
    File.open(infile, 'rb') do |input|
      Zlib::GzipWriter.open(outfile, 9) do |output|
        while data = input.read(1048576) do
          output.write(data)
        end
      end
    end
    
  rescue => e
    raise Crackup::CompressionError, "Unable to compress #{infile}: #{e}"
  end
  
  # Prints <em>message</em> to stdout if verbose mode is enabled.
  def self.debug(message)
    puts message if @options[:verbose] || $VERBOSE
  end
  
  # Reads <em>infile</em> and decompresses it to <em>outfile</em> using zlib
  # compression.
  def self.decompress_file(infile, outfile)
    Zlib::GzipReader.open(infile) do |input|
      File.open(outfile, 'wb') do |output|
        while data = input.read(1048576) do
          output.write(data)
        end
      end
    end
  
  rescue => e
    raise Crackup::CompressionError, "Unable to decompress #{infile}: #{e}"
  end
  
  # Calls GPG to decrypt <em>infile</em> to <em>outfile</em>.
  def self.decrypt_file(infile, outfile)
    File.delete(outfile) if File.exist?(outfile)

    gpg_command = String.new(GPG_DECRYPT)
    gpg_command.gsub!(':input_file', escapeshellarg(infile))
    gpg_command.gsub!(':output_file', escapeshellarg(outfile))
    gpg_command.gsub!(':passphrase', escapeshellarg(@options[:passphrase]))
    
    unless system(gpg_command)
      raise Crackup::EncryptionError, "Unable to decrypt file: #{infile}"
    end
  end
  
  def self.driver
    return @driver
  end
  
  # Calls GPG to encrypt <em>infile</em> to <em>outfile</em>.
  def self.encrypt_file(infile, outfile)
    File.delete(outfile) if File.exist?(outfile)

    gpg_command = String.new(GPG_ENCRYPT)
    gpg_command.gsub!(':input_file', escapeshellarg(infile))
    gpg_command.gsub!(':output_file', escapeshellarg(outfile))
    gpg_command.gsub!(':passphrase', escapeshellarg(@options[:passphrase]))
    
    unless system(gpg_command)
      raise Crackup::EncryptionError, "Unable to encrypt file: #{infile}"
    end
  end
  
  # Prints the specified <em>message</em> to stderr and exits with an error
  # code of 1.
  def self.error(message)
    abort "#{APP_NAME}: #{message}"
  end
  
  # Wraps <em>arg</em> in single quotes, escaping any single quotes contained
  # therein, thus making it safe for use as a shell argument.
  def self.escapeshellarg(arg)
    return "'#{arg.gsub("'", "\\'")}'"
  end
  
  # Gets an array of files in the remote file index whose local filenames match
  # <em>pattern</em>.
  def self.find_remote_files(pattern)
    files = []
    pattern.chomp!('/')
    
    @remote_files.each do |name, file|
      if File.fnmatch?(pattern, file.name)
        files << file
        next
      end

      next unless file.is_a?(Crackup::DirectoryObject)

      files += file.find(pattern)
    end
    
    return files
  end
  
  # Gets an array of filenames from <em>files</em>, which may be either a Hash
  # or a CrackupFileSystemObject.
  def self.get_list(files)
    list = []
  
    if files.is_a?(Hash)
      files.each_value {|value| list += get_list(value) }
    elsif files.is_a?(Crackup::DirectoryObject)
      list += get_list(files.children)
    elsif files.is_a?(Crackup::FileObject)
      list << files.name 
    end
    
    return list
  end

  # Gets a Hash of CrackupFileSystemObjects representing the files and
  # directories on the local system in the locations specified by the array of
  # filenames in <tt>options[:from]</tt>.
  def self.get_local_files
    local_files = {}
    
    @options[:from].each do |filename|
      next unless File.exist?(filename = filename.chomp('/'))
      next if local_files.has_key?(filename)
      
      # Skip this file if it's in the exclusion list.
      unless @options[:exclude].nil?
        next if @options[:exclude].any? do |pattern|
          File.fnmatch?(pattern, filename)
        end
      end
      
      if File.directory?(filename)
        debug "--> #{filename}"
        local_files[filename] = Crackup::DirectoryObject.from_path(filename)
      elsif File.file?(filename)
        debug "--> #{filename}"
        local_files[filename] = Crackup::FileObject.from_path(filename)
      end
    end
    
    return local_files
  end
  
  # Gets an array of CrackupFileSystemObjects representing files and directories
  # that exist at the remote location but no longer exist at the local location.
  def self.get_removed_files(local_files, remote_files)
    removed = []

    remote_files.each do |name, remotefile|
      unless local_files.has_key?(name)
        removed << remotefile
        next
      end

      localfile = local_files[name]
      
      if remotefile.is_a?(Crackup::DirectoryObject) && 
          localfile.is_a?(Crackup::DirectoryObject)
        removed += get_removed_files(localfile.children, remotefile.children)
      end
    end
    
    return removed
  end
  
  # Creates a new temporary file in the system's temporary directory and returns
  # its name. All temporary files will be deleted when the program exits.
  def self.get_tempfile
    tempfile = Tempfile.new('.crackup')
    tempfile.close
    
    return tempfile.path
  end

  # Gets an array of Crackup::FileSystemObjects representing files and
  # directories that are new or have been modified at the local location and
  # need to be updated at the remote location.
  def self.get_updated_files(local_files, remote_index)
    updated = []
    
    local_files.each do |name, localfile|
      # Add the file to the list if it doesn't exist at the remote location.
      unless remote_index.has_key?(name)
        updated << localfile
        next
      end
      
      remotefile = remote_index[name]
      
      if localfile.is_a?(Crackup::DirectoryObject) && 
          remotefile.is_a?(Crackup::DirectoryObject)
        # Add to the list all updated files contained in the directory and its 
        # subdirectories.
        updated += get_updated_files(localfile.children, remotefile.children)
      elsif localfile.is_a?(Crackup::FileObject) && 
          remotefile.is_a?(Crackup::FileObject)
        # Add the file to the list if the local file has been modified.
        unless localfile.file_hash == remotefile.file_hash
          updated << localfile
        end
      end
    end
    
    return updated
  end
  
  def self.options
    return @options
  end

  # Prints <em>message</em> to stdout and waits for user input, which is
  # then returned.
  def self.prompt(message)
    puts message + ': '
    return $stdin.gets
  end
  
  # Deletes each CrackupFileSystemObject specified in the <em>files</em> array
  # from the remote location.
  def self.remove_files(files)
    files.each do |file|
      file.remove
    end
  end
  
  # Uploads each CrackupFileSystemObject specified in the <em>files</em> array
  # to the remote location.
  def self.update_files(files)
    files.each do |file|
      file.update
    end
  end
  
end
